ntbs-service/DataMigration/NotificationMapper.cs

```diff
@@ -271,23 +256,53 @@ private async Task<IEnumerable<IList<Notification>>> GetGroupedResultsAsNotifica
             treatmentEvents.AddRange(notificationOutcomeEvents);
 
             // If there is a death date but no death treatment event, add one in
-            if (rawNotification.DeathDate != null
-                && !treatmentEvents.Any(e => _diedOutcomeIds.Contains(e.TreatmentOutcomeId ?? 0)))
+            if (rawNotification.DeathDate != null && !treatmentEvents.Any(e => e.TreatmentEventIsDeathEvent))
             {
-                treatmentEvents.Add(CreateDerivedDeathEvent(rawNotification));
+                treatmentEvents.Add(await CreateDerivedDeathEvent(rawNotification));
             }
 
             return treatmentEvents;
         }
 
-        private TreatmentEvent CreateDerivedDeathEvent(MigrationDbNotification rawNotification) =>
-            new TreatmentEvent
+        private void AlterStillOnTreatmentEventDateIfEndingOutcomePresent(Notification notification)
+        {
+            var lastEndingEvent = notification.TreatmentEvents.Where(te => te.TreatmentEventIsEndingEvent)
+                .GetMostRecentTreatmentEvent();
+            if (!notification.ClinicalDetails.StartingDate.HasValue || lastEndingEvent == null || !lastEndingEvent.EventDate.HasValue)
             {
-                EventDate = rawNotification.DeathDate,
-                TreatmentEventType = TreatmentEventType.TreatmentOutcome,
-                TreatmentOutcomeId = _postMortemOutcomeType.TreatmentOutcomeId,
-                TreatmentOutcome = _postMortemOutcomeType
-            };
+                return;
+            }
+
+            var expectedEndingOutcomeYear = GetExpectedEndingOutcomeYear(notification, lastEndingEvent);
+            var stillOnTreatmentEvent = notification.TreatmentEvents.OrderForEpisodes().LastOrDefault(o => o.TreatmentOutcome?.TreatmentOutcomeSubType == TreatmentOutcomeSubType.StillOnTreatment);
+
+            if (TreatmentOutcomesHelper.GetTreatmentOutcomeAtXYears(notification, expectedEndingOutcomeYear)?.TreatmentOutcomeId != lastEndingEvent.TreatmentOutcomeId
+                && stillOnTreatmentEvent != null)
+            {
+                var additionalNote = "Outcome date adjusted from initial date calculated in legacy system.";
+                stillOnTreatmentEvent.EventDate = lastEndingEvent.EventDate.Value.AddDays(-1);
+                stillOnTreatmentEvent.Note += additionalNote;
+            }
+        }
+
+        private int GetExpectedEndingOutcomeYear(Notification notification, TreatmentEvent endingEvent)
+        {
+            return new []{1, 2, 3}
+                .Where(n => notification.ClinicalDetails.StartingDate.Value.AddYears(n).AddDays(-1) > endingEvent.EventDate)
+                .DefaultIfEmpty(3).Min();
```

Comment:
Why is the default 3 here? If the ending event is more than 3 years after the start then I think this will be wrong. In reality if this is greater than 3 then we don't want to adjust any events.